tensorflow/core/kernels/tensor_forest/resources.cc

```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
   decision_tree_ = protobuf::Arena::CreateMessage<boosted_trees::Tree>(&arena_);
 }
 
+void TensorForestTreeResource::SplitNode(const int32 node,
+                                         boosted_trees::Vector& left_leaf,
+                                         boosted_trees::Vector& right_leaf,
+                                         std::vector<int32>* new_children) {
+  using boosted_trees::Leaf;
+  using boosted_trees::Node;
+  auto current_split =
+      decision_tree_->mutable_nodes(node)->mutable_dense_split();
+  current_split->Swap(best->mutable_split());
+  // add left leaf
+  auto* new_left = decision_tree_->add_nodes();
+  new_left->mutable_leaf()->mutable_vector()->Swap(left_leaf);
+  new_children->push_back(decision_tree_->nodes_size());
+  // add right leaf
+  auto* new_right = decision_tree_->add_nodes();
+  new_right->mutable_leaf()->mutable_vector()->Swap(right_leaf);
+  new_children->push_back(decision_tree_->nodes_size());
+};
+
+TensorForestFertileStatsResource::TensorForestFertileStatsResource()
+    : fertile_stats_(
+          protobuf::Arena::CreateMessage<tensor_forest::FertileStats>(
+              &arena_)){};
+
+bool TensorForestFertileStatsResource::InitFromSerialized(
+    const string& serialized) {
+  return ParseProtoUnlimited(fertile_stats_, serialized);
+};
+
+const tensor_forest::FertileStats&
+TensorForestFertileStatsResource::fertile_stats() const {
+  return *fertile_stats_;
+};
+
+void TensorForestFertileStatsResource::Reset() {
+  arena_.Reset();
+  CHECK_EQ(0, arena_.SpaceAllocated());
+  fertile_stats_ =
+      protobuf::Arena::CreateMessage<tensor_forest::FertileStats>(&arena_);
+};
+
+const bool TensorForestFertileStatsResource::IsSlotInitialized(
+    const int32 node_id, const int32 splits_to_consider) const {
+  if (fertile_stats_->node_to_slot().count(node_id) > 0) {
+    auto slot = fertile_stats_->node_to_slot().at(node_id);
+    return slot.leaf_stats().weight_sum() > 0 &&
+           slot.candidates_size() > splits_to_consider;
+  } else {
+    return false;
+  }
+};
+
+const bool TensorForestFertileStatsResource::IsSlotFinished(
+    const int32 node_id, const int32 split_nodes_after_samples,
+    const int32 splits_to_consider) const {
+  if (IsSlotInitialized(node_id, splits_to_consider)) {
+    auto slot = fertile_stats_->node_to_slot().at(node_id);
+    return slot.leaf_stats().weight_sum() > split_nodes_after_samples;
+  }
+  return false;
+};
+
+void TensorForestFertileStatsResource::UpdateSlotStats(
+    const bool is_regression, const int32 node_id, const int32 example_id,
```

Comment:
lets stick to node_id everywhere? in a previous file u call them node but it is still an int